File: 卡码/代码随想录/C++/单调栈/monotonicStack/LeetCode_0084.cpp
//84. 柱状图中最大的矩形
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
	int largestRectangleArea(vector<int>& heights) {
		heights.insert(heights.begin(),0);
		heights.push_back(0);
		stack<int> st;
		int res=0;
		for(int i=0;i<heights.size();i++){
			while(!st.empty()){
				if(heights[i]==heights[st.top()]) st.pop();
				else if(heights[i]<heights[st.top()]){
					int mid=st.top();
					st.pop();
					if(!st.empty()){
						int h=heights[mid];
						int w=i-st.top()-1;
						res=max(res,h*w);
					}
					continue;
				}
				break;
			}
			st.push(i);
		}
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/1/10-单调栈/Python/LeetCode_0084.py
#84. 柱状图中最大的矩形
from typing import *

class Solution:
	def largestRectangleArea(self, heights: List[int]) -> int:
		res:int=0
		heights.insert(0,0)
		heights.append(0)
		st:List[int]=[0]
		for i in range(1,len(heights)):
			if heights[i]==heights[st[-1]]:
				st.pop()
			else:
				while heights[i]<heights[st[-1]]:
					mid:int=st.pop()
					if len(st):
						h:int=heights[mid]
						w:int=i-st[-1]-1
						res=max(res,h*w)
			st.append(i)
		return res